video/out/drm_prime.c

```diff
@@ -78,15 +81,64 @@ int drm_prime_create_framebuffer(struct mp_log *log, int fd, AVDRMFrameDescripto
    return -1;
 }
 
-void drm_prime_destroy_framebuffer(struct mp_log *log, int fd, struct  drm_prime_framebuffer *framebuffer)
+void drm_prime_destroy_framebuffer(struct mp_log *log, int fd,
+                                    struct drm_prime_framebuffer *framebuffer,
+                                    struct drm_prime_handle_refs *handle_refs)
 {
     if (framebuffer->fb_id)
         drmModeRmFB(fd, framebuffer->fb_id);
 
     for (int i = 0; i < AV_DRM_MAX_PLANES; i++) {
-        if (framebuffer->gem_handles[i])
-            drmIoctl(fd, DRM_IOCTL_GEM_CLOSE, &framebuffer->gem_handles[i]);
+        if (framebuffer->gem_handles[i]) {
+            drm_prime_remove_handle_ref(handle_refs,
+                                         framebuffer->gem_handles[i]);
+            if(!drm_prime_get_handle_ref_count(handle_refs, framebuffer->gem_handles[i])) {
+                drmIoctl(fd, DRM_IOCTL_GEM_CLOSE, &framebuffer->gem_handles[i]);
+            }
+        }
     }
 
     memset(framebuffer, 0, sizeof(*framebuffer));
 }
+
+void drm_prime_init_handle_ref_count(void *talloc_parent, struct drm_prime_handle_refs *handle_refs) {
+    //4 dumb buffer + (1 per plane * planes we keep) + 10 (extra safety)
+    uint16_t init_size = 4 + (AV_DRM_MAX_PLANES * 4) + 10;
+    handle_refs->handle_ref_count = talloc_size(talloc_parent, init_size * sizeof(uint32_t));
```

Comment:
Could use `talloc_array(parent, uint32_t, init_size)`.

Reply:
PS: also could skip this, since drm_prime_add_handle_ref will allocate any necessary memory.

Reply:
Like I said I don't know how the resizing of arrays work with that macro. But I think it is the same as realloc but only with wrappers to have the tree. So it copys the data back and forth. With the initial size I'm trying to reduce the overhead here. Just pre-alloc enough and then if we still reach the boundary we can resize. Maybe init_size can be reduced. Meaning the extra safety will fall away.

Reply:
No, it checks the size of the allocation, and only reallocates if it's too small. When it reallocates, it always preallocates double the memory.

The extra realloc avoidance with the init_size isn't going to matter in this case. It seems better to reduce the size of the code.

Reply:
Changed it so that talloc does all allocation without any init preallocation.

Reply:
Thanks. That seems good so far. There's still an unnecessary allocation in the init code, but if you don't want to make any more changes, we can just leave that.

Still two issues before this can be merged:

- indentation issues (general mpv coding styles)
- squashing the commits into one (keeping the useful commit message of the first commit)

Does anyone want to do this?

Reply:
If you tell me the identation problems I can fix them. I think I made some mistakes at function definition (?). I'm not really familiar with the rules don't really have an ide right now.

After that I can do a squash into one commit.

Reply:
Sure, I just wanted to avoid "ordering" you too much around.

The coding style is mostly here: https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md#code-formatting

Function declarations are supposed to have the ``{`` on the next line. You often forget the space in ``if (``, and sometimes you mis-indent lines by a space.

Reply:
could you take a look at it ? Is it better now ? The identaion of line breaked function definition is mixed up in code but I stook to not having on space more in next line.

Reply:
Pointed out the remaining space issues I spotted.

You still need to squash your commits into 1. It's a single change and it makes sense if it's 1 commit. We prefer it this way so we can simply click on the merge button, instead of having to squash and push manually. (Github's squash-merge is useless to us.)

In addition, we prefer if lines are wrapped on preferably 80 columns. Not sure if there were any remaining instances of exceeding 85 columns (as hard limit). Github doesn't display it. Not so tragic if one or two lines go above that anyway.

Reply:
So I think I fixed everything. I quashed into first commit. Now it looks clean.

Reply:
Yes, very nice. Thank you for your patience.